### DISCUSSION TITLE
OPenmedia Vault in Proxmox Container

### DISCUSSION THREAD

**Discussion by InfoALR** [tteck/Proxmox#1075] [Help Needed]
Hello,

First of all thanks a lot for your amazing scripts.

I've installed the Openmediavault with your script in e container. Everything works.

Now I have the problem, even when I added a new mount point to the container of an other hard disk, the disk does not show up in the openemediavault. Its a normal ext4 share. The partition with the installation on omv is also not visible under Storage  over de GUI of OMV

Do you have an idea how I can solve this problem.

Thanks a lot.
Frank

![Screenshot 2023-02-19 at 01-19-09 pve - Proxmox Virtual Environment](https://user-images.githubusercontent.com/30112127/219905316-69fd460e-f792-47e2-bb00-667061515091.png)

**Comment by romka777** [comment]
~~OMV by default works only with bare hdd, but you cant attach bare hdd to lxc (this can be done only for vm).
So you need to install omv-extras (https://omv-extras.org), then goto plugins and install sharerootfs plugin, this will allow you to make share from root filesystem folders.~~
This is not working.

The working method:
1. Create omv lxc **privileged** container by running corresponding script (reserve storage only for omv system)
2. Stop CT
3. Go to Proxmox shell and add mount point by gui, or by cli, it will be use for storage:
`pct set <CT_ID> -mp0 local-lvm:10,mp=/data` 
will create 10GB storage 
4. Find block device ids:
Run `lvdisplay`, find Logical volume like `/dev/pve/vm-<CT_ID>-disk-X`, check it "Block device", its two numbers.
![image](https://user-images.githubusercontent.com/929742/229834073-2d469059-4760-4d09-b1bf-7108586d87bb.png)

5. Create mount hook script:
`echo '#!/bin/sh' > /var/lib/lxc/<CT_ID>/mount-hook.sh`
`echo 'mknod -m 777 ${LXC_ROOTFS_MOUNT}/dev/sda1 b 253 9' >> /var/lib/lxc/<CT_ID>/mount-hook.sh` (Replace 253 9 with device id numbers you got on step 4)
`chmod a+x /var/lib/lxc/<CT_ID>/mount-hook.sh`
6. Add hook to LXC config file:
`echo 'lxc.hook.autodev: /var/lib/lxc/<CT_ID>/mount-hook.sh' >> /etc/pve/lxc/<CT_ID>.conf`
7. Start CT 
8. There is no storage disk available, its ok
9. Goto Storage -> File Systems, click "Mount existing file system" and add "/dev/sda1"
10. Enjoy

**Reply by tteck** [reply]
Should this be added to the OMV script?

**Reply by InfoALR** [reply]
Hi guys, thanks a lot for your answer. I'll try that. I think this would be usefull, if it should be menntioned somewhere @tteck :)

Thanks again :)

**Reply by Koliera** [reply]
> Should this be added to the OMV script?

Dang, if you could do that then it would be great. The lack of this functionality is the only reason OMV is still the only non-lcx part of my proxmox setup.

**Reply by burgsth** [reply]
Sharerootfs plugin is part of main omv, omv-extras is not needed.
Unfortunately it does only create a warning "No root filesystem has been found." while installing and do not work at all.

**Reply by sarmenat0r** [reply]
Did you ever get this to work? I get the same warning and issue when installing the plugin

**Reply by InfoALR** [reply]
I must confirm too - it does not work.

**Reply by AndrewHaluza** [reply]
Some bullet points missed from this config;

1 to get blocks use `lsblk` (from a CT, not PVE). Here I've got block of `/dev/sda` and `sda*`
2  my mount-hook.sh 
```
#!/bin/sh

mknod -m 777 ${LXC_ROOTFS_MOUNT}/dev/sda b 8 0
mknod -m 777 ${LXC_ROOTFS_MOUNT}/dev/sda1 b 8 1
mknod -m 777 ${LXC_ROOTFS_MOUNT}/dev/sda2 b 8 2
mknod -m 777 ${LXC_ROOTFS_MOUNT}/dev/sda3 b 8 3
```
3 I'm lazy so I just allow all blocks in CT kernel like
` lxc.cgroup2.devices.allow: b 8:* rwm`
And finally threy're visible in OVM and available to mount. But keep in mind proxmox disk couldn't be mounted, as it should be wiped first, so it should be not used

**Reply by firinggamers** [reply]
i get this error 
[ERROR] in line 34: exit code 0: while executing command apt-get -y --auto-remove --show-upgraded --allow-downgrades --allow-change-held-packages --no-install-recommends --option DPkg::Options::="--force-confdef" --option DPkg::Options::="--force-confold" install openmediavault-keyring openmediavault &> /dev/null 
while it was installing omv

**Comment by romka777** [comment]
I just try it, confirm, this is not working, im sorry.

**Comment by romka777** [comment]
The working method:
1. Create omv lxc container by running corresponding script (reserve storage only for omv system)
2. Stop CT
3. Go to Proxmox shell and add mount point by gui, or by cli:
`pct set <CT_ID> -mp0 local-lvm:10,mp=/data` 
will create 10GB storage 
4. Find block device ids:
Run `lvdisplay`, find Logical volume like `/dev/pve/vm-<CT_ID>-disk-X`, check it "Block device", its two numbers.
![image](https://user-images.githubusercontent.com/929742/229834073-2d469059-4760-4d09-b1bf-7108586d87bb.png)

5. Create mount hook script:
`echo '#!/bin/sh' > /var/lib/lxc/<CT_ID>/mount-hook.sh`
`echo 'mknod -m 777 ${LXC_ROOTFS_MOUNT}/dev/sda1 b 253 9' >> /var/lib/lxc/<CT_ID>/mount-hook.sh` (Replace 253 9 with device id numbers you got on step 4)
`chmod a+x /var/lib/lxc/<CT_ID>/mount-hook.sh`
6. Add hook to LXC config file:
`echo 'lxc.hook.autodev: /var/lib/lxc/<CT_ID>/mount-hook.sh' >> /etc/pve/lxc/<CT_ID>.conf`
7. Start CT 
8. There is no storage disk available, its ok
9. Goto Storage -> File Systems, click "Mount existing file system" and add "/dev/sda1"
10. Enjoy

**Reply by Koliera** [reply]
:D .... ok :D Any chance to get the script to give you an option to add your hdd when creating the LCX? :D

**Reply by millie75** [reply]
Doesn't seem to work when adding from another disk.

I created a mount point on the container pointed to the other disk in this case sdb1, and when running lvdisplay I see nothing for this new mount point. How do you see the block device info for this point?

Thanks

**Reply by domoclick** [reply]
how can i view my Block device ids if i have zfs storage?

**Reply by Frew777** [reply]
> how can i view my Block device ids if i have zfs storage?

`lsblk -o name,fstype,fssize,maj:min,model,serial`

**Reply by JinKanzaki** [reply]
My CT won't start back up again after following your steps. No idea why.

**Reply by kvmgz** [reply]
Sadly I could manage to work only in a VM.
I'm willing to try your method so I can save up on resources.
My question would be, how to share that inserted disk with other LXC's, like Plex, and other -arr LXC's?

**Reply by romka777** [reply]
attach disk to omv, create shares you need, run nfs/smb/rsync on omv, then mount any share to any other LXC's

**Reply by PaPa1ya** [reply]
Does this work with ZFS? it doesn't look like the list of lsblk -o  name,fstype,fssize,maj:min,model,serial changes after adding the container storage.

UPD: In general, I decided to make a VM to be able to backup and transfer by means of Proxmox.

For those who want a container, here is an instruction where it can be done on zfs:
https://kawsing.gitbook.io/opensystem/andoid-shou-ji/pomoxve/proxmox-ve-+-openmediavault-shi-yong-lxc

**Reply by romka777** [reply]
I faced a problem - every time after reboot block device numbers are changed. So i adapted mount-hook script, now it:

```
#!/bin/sh
blkid=`lvdisplay | grep "DvsGGV-9xpT-egkL-21eX-gCPZ-f3KD-U3a1U3" -A11 | grep "Block device" | sed "s/.*\s\([0-9]\+\)\:\([0-9]\+\).*/\\1 \\2/g"`

mknod -m 777 ${LXC_ROOTFS_MOUNT}/dev/sda1 b $blkid
```
"DvsGGV-9xpT-egkL-21eX-gCPZ-f3KD-U3a1U3" - is UUID of my disk, you can find it by lvdisplay output

**Reply by Ky0sHiR00** [reply]
Hi,

Thank you for the updated script. Unfortunately, I ran into issue when starting omv container:

`TASK ERROR: lvm name '100/vm-100-disk-2.raw' contains illegal characters`

**Comment by burgsth** [comment]
Thank you romka777, works great.

Container needs to be privileged to work!

**Comment by OsipovOV** [comment]
I do everything according to the instructions, omv does not see the disk

**Comment by romka777** [comment]
> My CT won't start back up again after following your steps. No idea why.

check mount-hook.sh script, may be you lose "${LXC_ROOTFS_MOUNT}" part

**Reply by bkarakashev** [reply]
@romka777  what is the ${LXC_ROOTFS_MOUNT}? Do I need to replace it with some value or just use it as it is? 
Is it a built-in variable or something? When I echo $LXC_ROOTFS_MOUNT nothing is returned.

**Reply by romka777** [reply]
it is a built-in var

**Reply by bkarakashev** [reply]
@romka777  thanks for confirming, I had to update the script a bit and change the "-A11 " to "-A13" so now it returns the blkid.
However, I think I am missing something because when I try to "mount an existing file system" all I get is this:
<img width="1277" alt="Screenshot 2024-03-07 at 7 22 44 pm" src="https://github.com/tteck/Proxmox/assets/6206509/b3b5386b-c072-496c-a7e6-493ddf810121">

```
#!/bin/sh
blkid=`lvdisplay | grep "ZHoUJv-9fdF-0GQe-WvJw-TCJN-8uDj-HTshwY" -A13 | grep "Block device" | sed "s/.*\s\([0-9]\+\)\:\([0-9]\+\).*/\\1 \\2/g"`

mknod -m 777 ${LXC_ROOTFS_MOUNT}/dev/sdc1 b $blkid
```
any idea what I am doing wrong? Am I supposed to set the  `${LXC_ROOTFS_MOUNT}/dev/sdc1` to the id of the disk that I have used in the mount step 3. from the instructions?
Just to clarify I am trying to use an external USB disk that I have mounted to the node and set up as a Directory.
The OMV LXC is deployed on the local proxy disk, but I am trying to add external usb disk that is mounted as a type Directory.

**Reply by romka777** [reply]
This method is about mount proxmox storage to OMV LXC. 
You need to add your usb storage disk in proxmox to existing LVM-Thin or create new one. Then create virtual disk (mount point) on it for OMV LXC - step 3 in my guide.
If you want passthrough raw usb disk, you should create VM, not LXC.

**Comment by AndrewHaluza** [comment]
> Sadly I could manage to work only in a VM. I'm willing to try your method so I can save up on resources. My question would be, how to share that inserted disk with other LXC's, like Plex, and other -arr LXC's?

use NFS to attach disk to proxmox, then you're able to use disk in way you want. Or you even can attach NFS directly to you're plex server, or whatever

**Comment by petarlaf** [comment]
I can't seem to get this to work either. CT won't start. I suspect its because I'm trying to mount a 2nd hdd (which is on a different volume group?):

Start Log:

()
run_buffer: 322 Script exited with status 1
lxc_setup: 4445 Failed to run autodev hooks
do_start: 1272 Failed to setup container "108"
sync_wait: 34 An error occurred in another process (expected sequence number 4)
__lxc_start: 2107 Failed to spawn container "108"
TASK ERROR: startup for container '108' failed

Disk I'm trying to mount:

--- Logical volume ---
  LV Path                /dev/backup/vm-108-disk-1
  LV Name                vm-108-disk-1
  VG Name                backup
  LV UUID                vQTqXJ-OKiJ-t6je-HfHP-ZQpU-DGCR-BdiwEh
  LV Write Access        read/write
  LV Creation host, time prox, 2023-11-19 11:34:31 +0100
  LV Pool name           backup
  LV Status              available
  # open                 0
  LV Size                <1.76 TiB
  Mapped size            0.06%
  Current LE             460800
  Segments               1
  Allocation             inherit
  Read ahead sectors     auto
  - currently set to     256
  Block device           253:21
  
  ----
  
  root@prox:~# cat /var/lib/lxc/108/mount-hook.sh 
#!/bin/sh
mknod -m 777 ${LXC_ROOTFS_MOUNT}/dev/sda1 b 253 21

-----

root@prox:~# cat /etc/pve/lxc/108.conf 
## OMV LXC
#  ### https%3A//tteck.github.io/Proxmox/
#  <a href='https%3A//ko-fi.com/D1D7EP4GF'><img src='https%3A//img.shields.io/badge/%E2%98%95-Buy me a coffee-red' /></a>
arch: amd64
cores: 2
features: keyctl=1,nesting=1
hostname: omv
memory: 1024
mp0: backup:vm-108-disk-1,mp=/data,size=1800G
net0: name=eth0,bridge=vmbr0,hwaddr=6A:25:9D:94:96:7E,ip=dhcp,type=veth
onboot: 1
ostype: debian
rootfs: backup:vm-108-disk-0,size=4G
swap: 512
tags: proxmox-helper-scripts
unprivileged: 1
lxc.hook.autodev: /var/lib/lxc/108/mount-hook.sh

**Comment by dsolva** [comment]
Is this any easier since passthrough to lxc was added in Proxmox 8.1?

**Comment by DubanFlp** [comment]
Hi,

I am trying to deploy the OMV with the default options and it is failing, I also tried with the advanced option but the same thing happens. 

![Proxmox](https://github.com/user-attachments/assets/a4af45da-86f9-4c70-a4d1-c917dc8f47fd)

**Reply by tteck** [reply]
Whatever you're doing to execute the script, it's not running the `omv-install.sh` script.

**Reply by DubanFlp** [reply]
I ran the script again and it worked. Do you have any documentation on how to make the mount point visible from the GUI. Or is it better to create a VM instead of CT?
